From 70471115533e8807df51ea1870d1ef3de45c01c3 Mon Sep 17 00:00:00 2001
Message-Id: <70471115533e8807df51ea1870d1ef3de45c01c3.1434058173.git.chang-joon.lee@intel.com>
In-Reply-To: <cb13d6abcebb263859514735a399f7f8b895c27a.1434058173.git.chang-joon.lee@intel.com>
References: <cb13d6abcebb263859514735a399f7f8b895c27a.1434058173.git.chang-joon.lee@intel.com>
From: Liu Changcheng <changcheng.liu@intel.com>
Date: Fri, 13 Feb 2015 15:14:14 +0800
Subject: [PATCH 2/6] FOR_UPSTREAM: [VPG] drm/i915: fix event memory leak when
 flip timeout

We may alloc vblank event in drm_mode_page_flip_ioctl and/or
drm_mode_setplane, but we forget to free them when handing flip timeout.

Issue: IMINAN-37410
Change-Id: I4e2c9d44ac14328fcb721fd2352590bd91428b3c
Signed-off-by: Chuansheng Liu <chuansheng.liu@intel.com>
---
 drivers/gpu/drm/i915/intel_display.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index 74ac3c6..a615a9d 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -4075,12 +4075,31 @@ void intel_crtc_wait_for_pending_flips(struct drm_crtc *crtc)
 		spin_unlock_irqrestore(&dev->event_lock, flags);
 
 		if (unpin_work) {
+			struct drm_pending_vblank_event *vblank_e =
+				unpin_work->event;
+			if (vblank_e) {
+				spin_lock_irqsave(&dev->event_lock, flags);
+				vblank_e->base.file_priv->event_space +=
+					sizeof(vblank_e->event);
+				spin_unlock_irqrestore(&dev->event_lock, flags);
+				kfree(vblank_e);
+			}
+
 			intel_unpin_work_fn(&unpin_work->work);
 			atomic_clear_mask(1 << intel_crtc->plane,
 					&obj->pending_flip.counter);
 		}
 
 		if (sprite_unpin_work) {
+			struct drm_pending_vblank_event *vblank_e =
+				sprite_unpin_work->event;
+			if (vblank_e) {
+				spin_lock_irqsave(&dev->event_lock, flags);
+				vblank_e->base.file_priv->event_space +=
+					sizeof(vblank_e->event);
+				spin_unlock_irqrestore(&dev->event_lock, flags);
+				kfree(vblank_e);
+			}
 			obj = sprite_unpin_work->old_fb_obj;
 			if (obj)
 				atomic_clear_mask(1 << intel_crtc->plane,
-- 
1.9.1

